### ISSUE TITLE
MEM limited for training data

### ISSUE THREAD

**Issue by aldll** [lyst/lightfm#389]
Hi, 

I have a question about training data size.

My data contains three matrix: user-item interactions, user features, item features,
and if my item features is a very large matrix so that I can not load it limited by the memory size.

Is there any solution to deal with it?

Or I can train it separately by method .fit_partial() like this format?

suppose I have 10 item, and item features' dim is 5 (values here are just randomly generated)
split the item_features to two part
item_features_part_1:
```
[[ 0.63,  0.63,  0.09,  0.08,  0.49],
 [ 0.33,  0.22,  0.25,  0.66,  0.71],
 [ 0.7 ,  0.45,  0.03,  0.89,  0.21],
 [ 0.86,  0.17,  0.79,  0.64,  0.06],
 [ 0.23,  0.54,  0.44,  0.53,  0.76],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ]]
```

item_features_part_2:
```
[[ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
 [ 0.15,  0.73,  0.9 ,  0.2 ,  0.04],
 [ 0.33,  0.71,  0.13,  0.67,  0.26],
 [ 0.47,  0.65,  0.7 ,  0.01,  0.1 ],
 [ 0.46,  0.41,  0.41,  0.7 ,  0.22],
 [ 0.29,  0.29,  0.09,  0.96,  0.95]]
```
and do the same operation to interactions

`model = LightFM(loss='warp', no_components=100, random_state=0)`
`model.fit_partial(interactions_part_1, user_features=user_features_all, item_features=item_features_part_1, epochs=100)`

then del item_features_part_1 and load interactions_part2 & item_features_part_2

`model.fit_partial(interactions_part_2, user_features=user_features_all, item_features=item_features_part_2, epochs=100)`

Does this process do the same thing as training with the whole interactions and item features?

Thank you!

**Comment by maciejkula** [comment]
Are you using sparse matrices for your features? Would sparse matrices help?

Do the matrices themselves take a lot of memory, or is memory mostly used by various intermediate objects constructed as you are preparing your matrices?

**Comment by maciejkula** [comment]
For a back-of-the-envelope calculation, consider that a float matrix with 10 features and 25 million rows takes 1 gigabyte of memory.

**Comment by aldll** [comment]
Yes, I'm using sparse matrices. I simply load input data by method in lib scipy.sparse: `sp.load_npz()` and observe the memory usage.

Because a part of my item features matrix contains dense structure,
for example, an item has 300 features and sparse vector in first 200 features, the rest are dense vector,
so a float matrix with 300 features and 2 hundred rows takes 3 gigabyte of memory.

Or does it have other approaches to handle this kind of input data?

Thank you for your reply!

**Comment by maciejkula** [comment]
With LightFM, you can stream interactions data, but you always have to load the feature matrices into memory. You could experiment with using memory-mapped numpy arrays, but my recommendations would be to rent a bigger server.

**Comment by aldll** [comment]
Got it.

**Comment by maciejkula** [comment]
You can use memmapped arrays even when you are modifying them.

**Comment by aldll** [comment]
Yes, but memmapped can only be used in numpy array, not in sparse array.
and input data for model should use sparse array only. Is that right?

**Comment by maciejkula** [comment]
Sparse arrays are composed of constituent numpy arrays. You can memory map those.

**Comment by SimonCW** [comment]
I’m closing this issue because it has been inactive for a long time. If you still encounter the problem, please open a new issue.

Thank you!
